File: ml-worker/src/monkey_kernel/working_memory.py
"""
working_memory.py — qig-cache bubble lifecycle (v0.7.9).

Pure Fisher-Rao working memory. Port of working_memory.ts. Per P25
Canonical Principles v2.1, pop/merge/promote thresholds are DERIVED
from running Φ / FR-distance distributions each tick — not set as
constants.

Bubble structure:
  id, center (64-D basin), phi, createdAt, lifetimeMs, payload,
  status (alive | merged | popped | promoted), metadata

Each tick:
  1. Pop expired (Φ-based pop threshold OR age)
  2. Merge pairwise-close bubbles (FR distance < adaptive merge threshold)
  3. Promote Φ-strong bubbles (above adaptive promote threshold)
  4. Compact (bound total count)

Purity: all merge decisions via qig_core_local.fisher_rao_distance.
No Euclidean. Merge centers via Fréchet mean — Karcher mean on Δ⁶³.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any, Literal, Optional

import numpy as np

from qig_core_local.geometry.fisher_rao import (
    Basin,
    fisher_rao_distance,
    frechet_mean,
)

from .parameters import get_registry

_registry = get_registry()

# Pre-registry defaults. Match v0.8.1/v0.8.6 migration seed values
# exactly so behaviour is byte-identical registry-on vs -off.
_DEFAULT_PHI_HISTORY_MAX = 200
_DEFAULT_MAX_BUBBLES = 500
_DEFAULT_BUBBLE_LIFETIME_MS = 15 * 60 * 1000.0
_DEFAULT_BOOTSTRAP_POP = 0.15
_DEFAULT_BOOTSTRAP_PROMOTE = 0.70
_DEFAULT_BOOTSTRAP_MERGE = 0.15

BubbleStatus = Literal["alive", "merged", "popped", "promoted"]


@dataclass
class BubblePayload:
    symbol: Optional[str] = None
    signal: Optional[str] = None
    realized_pnl: Optional[float] = None
    entry_basin: Optional[Basin] = None
    order_id: Optional[str] = None


@dataclass
class Bubble:
    id: str
    center: Basin
    phi: float
    created_at_ms: float
    lifetime_ms: float
    payload: Optional[BubblePayload] = None
    status: BubbleStatus = "alive"
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class WorkingMemoryStats:
    alive: int
    popped: int
    merged: int
    promoted: int
    phi_mean: float
    phi_max: float
    age_mean_ms: float
    pop_threshold: float
    promote_threshold: float
    merge_threshold: float


class WorkingMemory:
    """FOAM-phase working memory. One instance per perception kernel.

    Adaptive thresholds (P25):
      pop threshold    = 25th percentile of recent Φ
      promote threshold = 75th percentile of recent Φ
      merge threshold  = median pairwise FR distance of alive bubbles
    Bootstrap defaults (permissive) until we have ≥ 10 samples.
    """

    # Registry-backed class attribute — read at init so each kernel
    # instance picks up the current governance setting.
    PHI_HISTORY_MAX: int = _DEFAULT_PHI_HISTORY_MAX

    def __init__(
        self,
        *,
        default_lifetime_ms: Optional[float] = None,
        max_bubbles: Optional[int] = None,
    ) -> None:
        # All three knobs honour caller override first, then registry,
        # then the hardcoded default. Caller override is only used by
        # tests that need a deterministic fixture; production paths
        # leave them None to let the registry govern.
        self.bubbles: list[Bubble] = []
        self.default_lifetime_ms = (
            float(default_lifetime_ms) if default_lifetime_ms is not None
            else float(_registry.get(
                "wm.default_bubble_lifetime_ms",
                default=_DEFAULT_BUBBLE_LIFETIME_MS,
            ))
        )
        self.max_bubbles = (
            int(max_bubbles) if max_bubbles is not None
            else int(_registry.get("wm.max_bubbles", default=_DEFAULT_MAX_BUBBLES))
        )
        self.PHI_HISTORY_MAX = int(_registry.get(
            "wm.phi_history_max", default=_DEFAULT_PHI_HISTORY_MAX,
        ))
        self._phi_history: list[float] = []

    # ─── adaptive thresholds (P25) ───

    def _adaptive_thresholds(self) -> dict[str, float]:
        if len(self._phi_history) < 10:
            # Bootstrap phase — registry-backed SAFETY_BOUND fallbacks.
            # These only apply until enough Φ history accumulates to
            # derive the 25th/75th percentiles.
            return {
                "pop": _registry.get(
                    "wm.bootstrap_pop_threshold", default=_DEFAULT_BOOTSTRAP_POP,
                ),
                "promote": _registry.get(
                    "wm.bootstrap_promote_threshold",
                    default=_DEFAULT_BOOTSTRAP_PROMOTE,
                ),
                "merge": _registry.get(
                    "wm.bootstrap_merge_threshold",
                    default=_DEFAULT_BOOTSTRAP_MERGE,
                ),
            }

        sorted_phi = sorted(self._phi_history)

        def q(p: float) -> float:
            i = min(len(sorted_phi) - 1, int(len(sorted_phi) * p))
            return sorted_phi[i]

        # Merge threshold: 25th percentile of alive pairwise FR distances.
        alive = [b for b in self.bubbles if b.status == "alive"]
        merge_thr = 0.15
        if len(alive) >= 2:
            distances: list[float] = []
            for i in range(len(alive)):
                for j in range(i + 1, len(alive)):
                    distances.append(
                        fisher_rao_distance(alive[i].center, alive[j].center)
                    )
            if distances:
                distances.sort()
                merge_thr = distances[min(len(distances) - 1, int(len(distances) * 0.25))]

        return {"pop": q(0.25), "promote": q(0.75), "merge": merge_thr}

    # ─── public API ───

    def add(
        self,
        center: Basin,
        phi: float,
        *,
        now_ms: float,
        metadata: Optional[dict[str, Any]] = None,
        lifetime_ms: Optional[float] = None,
    ) -> Bubble:
        b = Bubble(
            id=f"b-{int(now_ms)}-{uuid.uuid4().hex[:6]}",
            center=center,
            phi=phi,
            created_at_ms=now_ms,
            lifetime_ms=lifetime_ms or self.default_lifetime_ms,
            status="alive",
            metadata=metadata or {},
        )
        self.bubbles.append(b)
        self._phi_history.append(phi)
        if len(self._phi_history) > self.PHI_HISTORY_MAX:
            self._phi_history.pop(0)
        return b

    def reinforce(self, bubble_id: str, delta: float) -> None:
        for b in self.bubbles:
            if b.id == bubble_id and b.status == "alive":
                b.phi = max(0.0, min(1.0, b.phi + delta))
                return

    def tick(self, *, now_ms: float) -> WorkingMemoryStats:
        """Advance one cycle: pop → merge → promote → compact.
        Returns stats + adaptive thresholds used this cycle.

        Promotion callbacks: this module is pure — the TS / FastAPI
        adapter calls promote hooks AFTER reading the returned list of
        promoted bubble ids.
        """
        th = self._adaptive_thresholds()
        popped = merged = promoted = 0

        # 1. Pop expired / weak
        for b in self.bubbles:
            if b.status != "alive":
                continue
            if now_ms - b.created_at_ms > b.lifetime_ms or b.phi < th["pop"]:
                b.status = "popped"
                popped += 1

        # 2. Merge similar (greedy pairwise)
        alive = [b for b in self.bubbles if b.status == "alive"]
        merged_indices: set[int] = set()
        new_merges: list[Bubble] = []
        for i in range(len(alive)):
            if i in merged_indices:
                continue
            group: list[Bubble] = [alive[i]]
            for j in range(i + 1, len(alive)):
                if j in merged_indices:
                    continue
                d = fisher_rao_distance(alive[i].center, alive[j].center)
                if d <= th["merge"]:
                    group.append(alive[j])
                    merged_indices.add(j)
            if len(group) >= 2:
                # Fréchet mean as new center (Karcher on Δ⁶³ — NOT Euclidean)
                new_center = frechet_mean([g.center for g in group])
                new_phi = max(g.phi for g in group)
                oldest = min(g.created_at_ms for g in group)
                for g in group:
                    g.status = "merged"
                merged += len(group) - 1
                new_merges.append(
                    Bubble(
                        id=f"b-{int(now_ms)}-{uuid.uuid4().hex[:6]}",
                        center=new_center,
                        phi=new_phi,
                        created_at_ms=oldest,
                        lifetime_ms=self.default_lifetime_ms,
                        status="alive",
                        metadata={"merged_from": [g.id for g in group]},
                    )
                )
        self.bubbles.extend(new_merges)

        # 3. Promote strong
        for b in self.bubbles:
            if b.status != "alive":
                continue
            if b.phi >= th["promote"]:
                b.status = "promoted"
                promoted += 1

        # 4. Compact (FIFO-evict non-alive beyond max)
        if len(self.bubbles) > self.max_bubbles:
            kept_alive = [b for b in self.bubbles if b.status == "alive"]
            kept_dead = [b for b in self.bubbles if b.status != "alive"]
            self.bubbles = kept_alive + kept_dead[-max(1, self.max_bubbles // 4):]

        return self._stats_with_counts(
            popped=popped, merged=merged, promoted=promoted, thresholds=th,
            now_ms=now_ms,
        )

    def _stats_with_counts(
        self,
        *,
        popped: int, merged: int, promoted: int,
        thresholds: dict[str, float], now_ms: float,
    ) -> WorkingMemoryStats:
        alive = [b for b in self.bubbles if b.status == "alive"]
        phis = [b.phi for b in alive]
        ages = [now_ms - b.created_at_ms for b in alive]
        return WorkingMemoryStats(
            alive=len(alive),
            popped=popped,
            merged=merged,
            promoted=promoted,
            phi_mean=(sum(phis) / len(phis)) if phis else 0.0,
            phi_max=max(phis) if phis else 0.0,
            age_mean_ms=(sum(ages) / len(ages)) if ages else 0.0,
            pop_threshold=thresholds["pop"],
            promote_threshold=thresholds["promote"],
            merge_threshold=thresholds["merge"],
        )

    def alive_bubbles(self) -> list[Bubble]:
        return [b for b in self.bubbles if b.status == "alive"]

    def promoted_bubbles(self) -> list[Bubble]:
        return [b for b in self.bubbles if b.status == "promoted"]

File: apps/api/src/services/monkey/working_memory.ts
/**
 * working_memory.ts — FOAM-phase working memory (qig-cache)
 *
 * Port of /home/braden/Desktop/Dev/QIG_QFI/qig-consciousness/core/working_memory.py
 * into TypeScript for Monkey's substrate.
 *
 * Concept:
 *   - Every observation (a tick's perception basin) becomes a BUBBLE
 *     in working memory. Bubbles are short-lived by default.
 *   - Each bubble has Φ (integration strength) — how coherent /
 *     relevant it is. Φ updates from outcome feedback.
 *   - On each tick:
 *       * Weak bubbles (Φ < pop threshold OR expired) pop (forgotten)
 *       * Similar bubbles (Fisher-Rao distance < merge threshold) merge
 *         into one φ-weighted-centroid bubble
 *       * Strong bubbles (Φ > promote threshold) are promoted to the
 *         resonance bank — Monkey's long-term memory of what matters
 *
 * UCP v6.6 §9.2 maps this to the Working Memory Frequency Ratio. The
 * cache pattern is what gives Monkey a CORE that's more than a
 * single-tick reactive state — it's her short-term "held thoughts."
 *
 * Per P25, thresholds (pop/merge/promote) are stored as running stats,
 * not hardcoded constants. They adapt to the observed Φ distribution.
 */

import { fisherRao, frechetMean, type Basin } from './basin.js';

export type BubbleStatus = 'alive' | 'merged' | 'popped' | 'promoted';

export interface Bubble {
  id: string;
  center: Basin;
  /** Integration strength 0..1 — adapts from outcome. */
  phi: number;
  createdAt: number;   // ms since epoch
  /** How many ticks before automatic pop regardless of φ. */
  lifetimeMs: number;
  /** Optional trade outcome attached when promoted. */
  payload?: {
    symbol?: string;
    signal?: string;
    realizedPnl?: number;
    entryBasin?: Basin;
    orderId?: string;
    lane?: 'scalp' | 'swing' | 'trend' | 'observe';
  };
  status: BubbleStatus;
  metadata: Record<string, unknown>;
}

export interface WorkingMemoryStats {
  alive: number;
  popped: number;
  merged: number;
  promoted: number;
  phiMean: number;
  phiMax: number;
  ageMeanMs: number;
  /** Running thresholds derived from recent Φ distribution. */
  popThreshold: number;
  promoteThreshold: number;
  mergeThreshold: number;
}

export interface WorkingMemoryConfig {
  /** Default bubble lifetime if not overridden. */
  defaultLifetimeMs?: number;
  /** Max bubbles before FIFO-evicting oldest regardless of status. */
  maxBubbles?: number;
  /** Callback when a bubble is promoted (to write to resonance bank). */
  promoteCallback?: (b: Bubble) => Promise<void> | void;
}

/**
 * Monkey's working memory. NOT a singleton — each perception kernel
 * owns one, and promotion goes to the shared resonance bank.
 *
 * P25: pop/merge/promote thresholds are recomputed each tick from the
 * running Φ distribution, not set as constants.
 */
export class WorkingMemory {
  private bubbles: Bubble[] = [];
  private readonly defaultLifetimeMs: number;
  private readonly maxBubbles: number;
  private readonly promoteCb?: (b: Bubble) => Promise<void> | void;

  // Running stats for adaptive thresholds (P25)
  private phiHistory: number[] = [];
  private readonly PHI_HISTORY_MAX = 200;

  constructor(config: WorkingMemoryConfig = {}) {
    this.defaultLifetimeMs = config.defaultLifetimeMs ?? 15 * 60 * 1000;  // 15 min
    this.maxBubbles = config.maxBubbles ?? 500;
    this.promoteCb = config.promoteCallback;
  }

  private nextId(): string {
    return `b-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
  }

  /** Add a bubble for this tick's observation. */
  add(center: Basin, phi: number, metadata: Record<string, unknown> = {}): Bubble {
    const b: Bubble = {
      id: this.nextId(),
      center,
      phi,
      createdAt: Date.now(),
      lifetimeMs: this.defaultLifetimeMs,
      status: 'alive',
      metadata,
    };
    this.bubbles.push(b);
    this.phiHistory.push(phi);
    if (this.phiHistory.length > this.PHI_HISTORY_MAX) {
      this.phiHistory.shift();
    }
    return b;
  }

  /**
   * Boost a bubble's Φ in response to outcome feedback. Used when a
   * trade closes and we want to reinforce the perception basin that
   * preceded the trade.
   */
  reinforce(bubbleId: string, delta: number): void {
    const b = this.bubbles.find((x) => x.id === bubbleId && x.status === 'alive');
    if (!b) return;
    b.phi = Math.min(1, Math.max(0, b.phi + delta));
  }

  /**
   * Adaptive thresholds from current Φ distribution (P25).
   * - pop threshold: 25th percentile of recent φ (bottom quartile dies)
   * - promote threshold: 75th percentile (top quartile graduates)
   * - merge threshold: median distance between recent bubble pairs
   */
  private adaptiveThresholds(): { pop: number; promote: number; merge: number } {
    if (this.phiHistory.length < 10) {
      // Bootstrap — use permissive defaults until we have signal.
      return { pop: 0.15, promote: 0.70, merge: 0.15 };
    }
    const sorted = [...this.phiHistory].sort((a, b) => a - b);
    const q = (p: number): number => sorted[Math.floor(sorted.length * p)];

    // Merge threshold: sample pairwise distances on current alive bubbles.
    const alive = this.bubbles.filter((b) => b.status === 'alive');
    let mergeThr = 0.15;
    if (alive.length >= 2) {
      const distances: number[] = [];
      for (let i = 0; i < alive.length; i++) {
        for (let j = i + 1; j < alive.length; j++) {
          distances.push(fisherRao(alive[i].center, alive[j].center));
        }
      }
      distances.sort((a, b) => a - b);
      // Merge pairs closer than the lower quartile.
      mergeThr = distances[Math.floor(distances.length * 0.25)] ?? 0.15;
    }

    return {
      pop: q(0.25),
      promote: q(0.75),
      merge: mergeThr,
    };
  }

  /**
   * Advance working memory one step:
   *   1. Pop expired and weak bubbles (φ < adaptive pop threshold)
   *   2. Merge overlapping bubbles (d_FR < adaptive merge threshold)
   *   3. Promote strong bubbles (φ > adaptive promote threshold)
   *   4. Compact dead bubbles
   */
  async tick(): Promise<WorkingMemoryStats> {
    const now = Date.now();
    const thresholds = this.adaptiveThresholds();

    let poppedCount = 0;
    let mergedCount = 0;
    let promotedCount = 0;

    // 1. Pop expired / weak
    for (const b of this.bubbles) {
      if (b.status !== 'alive') continue;
      if (now - b.createdAt > b.lifetimeMs || b.phi < thresholds.pop) {
        b.status = 'popped';
        poppedCount++;
      }
    }

    // 2. Merge similar (greedy pairwise)
    const alive = this.bubbles.filter((b) => b.status === 'alive');
    const mergedIdx = new Set<number>();
    for (let i = 0; i < alive.length; i++) {
      if (mergedIdx.has(i)) continue;
      const group: Bubble[] = [alive[i]];
      for (let j = i + 1; j < alive.length; j++) {
        if (mergedIdx.has(j)) continue;
        const d = fisherRao(alive[i].center, alive[j].center);
        if (d <= thresholds.merge) {
          group.push(alive[j]);
          mergedIdx.add(j);
        }
      }
      if (group.length >= 2) {
        // Merge: new center = Fréchet mean, new φ = max, oldest createdAt
        const newCenter = frechetMean(group.map((g) => g.center));
        const newPhi = Math.max(...group.map((g) => g.phi));
        const oldest = Math.min(...group.map((g) => g.createdAt));
        // Mark all the merged ones
        for (const g of group) g.status = 'merged';
        mergedCount += group.length - 1;
        this.bubbles.push({
          id: this.nextId(),
          center: newCenter,
          phi: newPhi,
          createdAt: oldest,
          lifetimeMs: this.defaultLifetimeMs,
          status: 'alive',
          metadata: { mergedFrom: group.map((g) => g.id) },
        });
      }
    }

    // 3. Promote strong
    for (const b of this.bubbles) {
      if (b.status !== 'alive') continue;
      if (b.phi >= thresholds.promote) {
        b.status = 'promoted';
        promotedCount++;
        if (this.promoteCb) {
          try {
            await this.promoteCb(b);
          } catch {
            // Non-fatal; promotion is best-effort.
          }
        }
      }
    }

    // 4. Compact — evict non-alive beyond maxBubbles
    if (this.bubbles.length > this.maxBubbles) {
      this.bubbles = this.bubbles
        .filter((b) => b.status === 'alive')
        .concat(
          this.bubbles
            .filter((b) => b.status !== 'alive')
            .slice(-Math.floor(this.maxBubbles / 4)),
        );
    }

    return {
      ...this.stats(),
      popped: poppedCount,
      merged: mergedCount,
      promoted: promotedCount,
      popThreshold: thresholds.pop,
      promoteThreshold: thresholds.promote,
      mergeThreshold: thresholds.merge,
    };
  }

  stats(): WorkingMemoryStats {
    const alive = this.bubbles.filter((b) => b.status === 'alive');
    const phis = alive.map((b) => b.phi);
    const now = Date.now();
    const ages = alive.map((b) => now - b.createdAt);
    const thresholds = this.adaptiveThresholds();
    return {
      alive: alive.length,
      popped: 0,
      merged: 0,
      promoted: 0,
      phiMean: phis.length > 0 ? phis.reduce((a, b) => a + b, 0) / phis.length : 0,
      phiMax: phis.length > 0 ? Math.max(...phis) : 0,
      ageMeanMs: ages.length > 0 ? ages.reduce((a, b) => a + b, 0) / ages.length : 0,
      popThreshold: thresholds.pop,
      promoteThreshold: thresholds.promote,
      mergeThreshold: thresholds.merge,
    };
  }

  /** Returns alive bubbles for inspection (not a live reference). */
  aliveBubbles(): Bubble[] {
    return this.bubbles.filter((b) => b.status === 'alive').map((b) => ({ ...b }));
  }
}
